### PR TITLE
[modules][Android] Fix modules aren't created after reload

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

A follow-up to https://github.com/expo/expo/pull/18844 and https://github.com/expo/expo/pull/18823.
It turns out that the recently added fix breaks the reload behavior. Right now, when you hard reload your app, modules won't be installed, because the `mModulesProxy` isn't recreated. 

# How

Made sure that the module proxy is recreated after reload.
Added some debug logs. 

> Note:
There are still some problems when someone reloads the app multiple times.  

# Test Plan

- unit tests ✅
- bare-expo ✅
- expo-go ✅

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt [hunk 1]
```diff
@@ -95,18 +97,25 @@ class AppContext(
...
+        ?.let {
+          jsiInterop.installJSI(
+            it,
+            catalystInstance.jsCallInvokerHolder as CallInvokerHolderImpl,
+            catalystInstance.nativeCallInvokerHolder as CallInvokerHolderImpl
+          )
+          Log.i("ExpoModulesCore", "✅ JSI interop was installed")
+        }
+    } catch (e: Throwable) {
+      Log.e("ExpoModulesCore", "Cannot install JSI interop: $e", e)rrrr
```
Comment (lukmccall) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:117]:
```suggestion
      Log.e("ExpoModulesCore", "Cannot install JSI interop: $e", e)
```

**Review by Kudo — APPROVED** [review]
do we want to have the successful log event after released?
e.g. `Log.i("ExpoModulesCore", "✅ Constants was exported");`

**Comment by lukmccall** [comment]
> do we want to have the successful log event after released? e.g. `Log.i("ExpoModulesCore", "✅ Constants was exported");`

I'm not sure but I would say yes. The release log level should be set to the info.

**Comment by Kudo** [comment]
> I'm not sure but I would say yes. The release log level should be set to the info.

since the log will appear in adb logcat even successful, i'm afraid if there're too much successful debug log.
currently we don't have much successful debug log though, it's fine for me to do so.
please keep as-is and thanks for this neat fix 👍

**Comment by touchmarine** [comment]
Here is a patch until this fix is released: https://gist.github.com/touchmarine/4332744e177656d139644b3fd7fc9eba

Tested on expo-modules-core@0.11.5 (Expo SDK 46)

**Pull request merged**